orangecontrib/prototypes/widgets/owexplpred.py

```diff
@@ -0,0 +1,401 @@
+                    task.watcher = FutureWatcher(task.future)
+                    task.watcher.done.connect(self._task_finished)
+            else:
+                self.Warning.selection_not_matching()
+        else:
+            self.Outputs.explanations.send(None)
+
+    @pyqtSlot(concurrent.futures.Future)
+    def _task_finished(self, f):
+        """
+        Parameters:
+        ----------
+        f: conncurent.futures.Future
+            future instance holding the result of learner evaluation
+        """
+        assert self.thread() is QThread.currentThread()
+        assert self._task is not None
+        assert self._task.future is f
+        assert f.done()
+
+        self._task = None
+
+        try:
+            results = f.result()
+        except Exception as ex:
+            log = logging.getLogger()
+            log.exception(__name__, exc_info=True)
+            self.error("Exception occured during evaluation: {!r}".format(ex))
+
+            for key in self.results.keys():
+                self.results[key] = None
+        else:
+            classValue = results[0]
+            self.explanations = results[1]
+            self._print_prediction(classValue)
+            self.Outputs.explanations.send(self.explanations)
+            model = TableModel(self.explanations, parent=None)
+            self.dataview.setModel(model)
+
+    def cancel(self):
+        """
+        Cancel the current task (if any).
+        """
+        if self._task is not None:
+            self._task.cancel()
+            assert self._task.future.done()
+            # disconnect the `_task_finished` slot
+            self._task.watcher.done.disconnect(self._task_finished)
+            self._task_finished(self._task.future)
+
+    def _print_prediction(self, classValue):
+        """
+        Parameters
+        ----------
+        classValue: float 
+            Number representing either index of predicted class value, looked up in domain, or predicted value (regression)
+        """
+        name = self.data.domain.class_vars[0].name
+        if isinstance(self.data.domain.class_vars[0], ContinuousVariable):
+            self.predict_info.setText(name + ":      " + str(classValue))
+        else:
+            self.predict_info.setText(
+                name + ":      " + self.data.domain.class_vars[0].values[int(classValue)])
+
+    def _update_error_spin(self):
+        self.cancel()
+        self.handleNewSignals()
+
+    def _update_p_val_spin(self):
+        self.cancel()
+        self.handleNewSignals()
+
+    def onDeleteWidget(self):
+        self.cancel()
+        super().onDeleteWidget()
+
+
+def domain_equal(domain1, domain2):
+    """checks if two domains have the same attributes and target values"""
+
+    if len(domain1.class_vars) != len(domain2.class_vars):
+        return False
+    for idx in range(len(domain1.class_vars)):
+        if domain1.class_vars[idx].name != domain2.class_vars[idx].name:
```

Comment:
a) why are you checking classes? isn't a sample without a class an acceptable instance to be explained
b) why are you comparing names? are variable comparisons not appropriate here for some reason?
c) why are you doing pairwise comparisons in a for loop? isn't `domain1.class_vars == domain2.class_vars` enough?
d) if you needed a loop, it would be nicer to iterate over pairs with zip: `for v1, v2 in zip(domain1.class_vars, domain2.class_vars)` try to avoid `range(len(...` when you do not need the actual indices and can use the elements directly instead

Reply:
This function is made because testing for equality  with == would mean the variable from one domain were copies of variables from the other. I was trying to account for some weird case where sample to be explained is not just a selected row from dataset that is the input for widget, but comes from another dataset, that still has the same domain. In that case, vars wouldn't be copies and == would return false.
If there won't be such a case, i'll change it to ==.

Reply:
Well, if you opened the same kind of file in Orange the Variable descriptors would most likely get reused.

As your code assumes that columns have the same meaning in both data and instance just looking at the name is not enough. You would definitely need equality.

I would try the same approach we use for predictions: convert into other domain. Like starting with (I assume that instance is Table with only one row) `instance = instance.transform(data.domain)` or vice versa. The values Orange will not know how to match will be set to unknown. This would also handle the case if data is somehow (pre)processed and instance is raw, unprocessed instance.

Reply:
Thank you Marko, seems like an elegant solution, will do that.